src/GafferArnold/IECoreArnoldPreview/VDBAlgo.cpp

```diff
@@ -0,0 +1,102 @@
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Image Engine Design nor the names of any
+//       other contributors to this software may be used to endorse or
+//       promote products derived from this software without specific prior
+//       written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECore/SimpleTypedData.h"
+#include "IECore/Renderer.h"
+#include "IECore/CompoundData.h"
+
+#include "IECoreArnold/ParameterAlgo.h"
+
+#include "IECoreArnold/NodeAlgo.h"
+
+#include "GafferVDB/VDBObject.h"
+
+using namespace std;
+using namespace Imath;
+
+using namespace IECore;
+using namespace IECoreArnold;
+
+//////////////////////////////////////////////////////////////////////////
+// Implementation of public API
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+	IECore::InternedString g_filedataParam("filedata");
+	IECore::InternedString g_gridParam("grids");
+	AtString g_volume("volume");
+
+	CompoundDataPtr  createParameters(const GafferVDB::VDBObject* vdbObject)
+	{
+		CompoundDataPtr parameters = new CompoundData();
+		CompoundDataMap& compoundData = parameters->writable();
+
+		compoundData[g_gridParam] = new StringVectorData( vdbObject->gridNames() );
+		compoundData[g_filedataParam] = vdbObject->memoryBuffer();
+
+		return parameters;
+	}
+}
+
+namespace IECoreArnoldPreview
+{
+
+namespace VDBAlgo
+{
+
+AtNode *convert( const GafferVDB::VDBObject *vdbObject, const std::string & /*name*/, const AtNode* /* parent */ )
+{
+	AtNode *node = AiNode( g_volume );
+
+	CompoundDataPtr parameters = createParameters( vdbObject );
+	ParameterAlgo::setParameters( node, parameters->readable() );
```

Comment:
Since the serialised VDB data could be huuge, can we reduce the number of copies we're making here? Ideally we'd stream straight into an AtArray, which I presume isn't possible, but can we at least avoid the intermediate `VDBObject::memoryBuffer()`? Actually, I think there's a good case for not providing `VDBObject::memoryBuffer()` at all, and treating VDBObject as a mere container, with everything else expected to be done either directly via openvdb methods, or via external utility functions.

Reply:
> Since the serialised VDB data could be huuge

Are we causing ourselves problems by always forcing VDBs to be completely loaded in memory before rendering? We could have a read from file path to make this the renderers problem if we don't change the VDB in memory?

> Ideally we'd stream straight into an AtArray, which I presume isn't possible

We can if we construct a byte array using AiArray and get the raw write pointer using AiArrayMap and use this to back the iostream sink device. The only problem is knowing the exact size of serialised VDB. 

I hope I've reduced the number of copies in https://github.com/GafferHQ/gaffer/pull/2328/commits/6e1ed096074a7fbac07219dabf46d67b3ce991ee by using a boost iostreams sink which writes to a preallocated UCharVectorData. 

> Actually, I think there's a good case for not providing VDBObject::memoryBuffer() at all, and treating VDBObject as a mere container, with everything else expected to be done either directly via openvdb methods, or via external utility functions.

I'm not sure I follow what you're suggesting here.

Reply:
> Are we causing ourselves problems by always forcing VDBs to be completely loaded in memory before rendering? We could have a read from file path to make this the renderers problem if we don't change the VDB in memory?

Good idea - it does seem like that could be a common enough occurrence that we might want to optimise for it. How would we detect it? The VDBObject was initialised from a file, and nobody has used a non-const grid accessor yet?

>> Actually, I think there's a good case for not providing VDBObject::memoryBuffer() at all, and treating VDBObject as a mere container, with everything else expected to be done either directly via openvdb methods, or via external utility functions.
> I'm not sure I follow what you're suggesting here.

I'm saying that VDBObject's primary purpose is to be a container of grids so that they can be passed through a Gaffer graph. It needn't provide anything beyond that - the VDB API itself could be used directly instead, or the utilities could exist elsewhere where they can be used on non-vdb-object-wrapped grids. In this specific case, perhaps `memoryBuffer()` could just be an internal utility in IECoreArnold, at least until it proves itself needed elsewhere. This is of much lower importance than sorting out the other outstanding issues though...